internal/resourcepolicies/resource_policies_mgr.go

```diff
@@ -0,0 +1,139 @@
+	namespace   string
+	kbClient    kbClient.Client
+}
+
+var ResPoliciesMgr ResPoliciesManager
+
+func (mgr *ResPoliciesManager) InitResPoliciesMgr(namespace string, kbClient kbClient.Client) {
+	mgr.resPolicies = make(map[resPoliciesKey]*resPoliciesMeta)
+	mgr.namespace = namespace
+	mgr.kbClient = kbClient
+}
+
+// getResourcePolicies first check and update current resource policies and then get the latest resource policies
+func (mgr *ResPoliciesManager) getResourcePolicies(refName, refType string) (*ResourcePolicies, error) {
+	if err := mgr.updateResourcePolicies(refName, refType); err != nil {
+		return nil, errors.Wrap(err, "failed to update resource policies meta info")
+	}
+
+	resKey := resPoliciesKey{
+		refName: refName,
+		refType: refType,
+	}
+	if policies, exist := mgr.resPolicies[resKey]; exist {
+		return policies.policies, nil
+	}
+	return nil, fmt.Errorf("failed to get resource policies")
+}
+
+func (mgr *ResPoliciesManager) updateResourcePolicies(refName, refType string) error {
+	// TODO currently we only want to support type of configmap
+	if refType != "configmap" {
+		return fmt.Errorf("unsupported resource policies reference type %s", refType)
+	}
+
+	resKey := resPoliciesKey{
+		refName: refName,
+		refType: refType,
+	}
+	policiesConfigmap := &v1.ConfigMap{}
+	err := mgr.kbClient.Get(context.Background(), kbClient.ObjectKey{Namespace: mgr.namespace, Name: refName}, policiesConfigmap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get resource policies %s/%s configmap", refName, mgr.namespace)
+	}
+	if policiesConfigmap == nil {
+		return errors.Wrapf(err, "get empty resource policies %s/%s configmap", refName, mgr.namespace)
+	}
+
+	updateResPolicies := func() error {
+		res, err := GetResourcePoliciesFromConfig(policiesConfigmap)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the user resource policies config")
+		}
+		if res == nil {
+			return fmt.Errorf("%v resource policy not found", resKey)
+		}
+		mgr.resPolicies[resKey] = &resPoliciesMeta{
+			policies:                res,
+			configCreationTimestamp: policiesConfigmap.CreationTimestamp,
+		}
+		return nil
+	}
+
+	if policies, exist := mgr.resPolicies[resKey]; exist {
+		// check if it needs update
+		if !policiesConfigmap.CreationTimestamp.Equal(&policies.configCreationTimestamp) {
+			if err := updateResPolicies(); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := updateResPolicies(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (mgr *ResPoliciesManager) GetStructredVolumeForPVC(namespace, name string) (*StructuredVolume, error) {
+	pvc := &v1.PersistentVolumeClaim{}
+	err := mgr.kbClient.Get(context.Background(), kbClient.ObjectKey{Namespace: namespace, Name: name}, pvc)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if pvc.Spec.VolumeName == "" {
+		return nil, errors.Errorf("PVC %s/%s has no volume backing this claim", namespace, name)
+	}
+	if pvc.Status.Phase != v1.ClaimBound {
```

Comment:
Should we return the error or just skip it?

Reply:
both ok. we could skip checking it here because it will check later in CSI plugin?